### PR TITLE
Add cats-effect support for any type with Async

### PR DESCRIPTION
**Pull request by andimiller**
This changes the cats-effect support from being hard coupled to `cats.effect.IO` over to any type which has `cats.effect.Async` implemented for it.

### CODE REVIEW

File: pulsar4s-cats-effect/src/main/scala/com/sksamuel/pulsar4s/cats/CatsAsyncHandler.scala [hunk 1]
```diff
@@ -3,66 +3,75 @@ package com.sksamuel.pulsar4s.cats
...
+            }
+          })
         }
-      })
+
     }
+  }
 
-  override def failed(e: Throwable): IO[Nothing] = IO.raiseError(e)
+  implicit def asyncHandlerForCatsEffectAsync[F[_]: Async]: AsyncHandler[F] = new AsyncHandler[F] {
```
Comment (gmethvin) [inline pulsar4s-cats-effect/src/main/scala/com/sksamuel/pulsar4s/cats/CatsAsyncHandler.scala]: Does this need to be `implicit`? It seems like it's always being explicitly called.

Comment (andimiller) [reply pulsar4s-cats-effect/src/main/scala/com/sksamuel/pulsar4s/cats/CatsAsyncHandler.scala]: this needs to be implicit to provide `AsyncHandler` instances for any `F[_]` that has an `Async`, you can see it's usage in the new tests I've added

**Comment by andimiller** [comment]
I've added examples written in a tagless final style, using `cats.effect.IO` a `cats.effect.IO` fed through `cats.data.StateT`, `monix.Task` and `zio.Task`

**Comment by andimiller** [comment]
This becomes a breaking change because you can't import the package to get a single `AsyncHandler` for `cats.effect.IO`, so this may belong in a separate object, I'm not sure what's more user friendly.

As a user not familiar with the project, I expected the `cats-effect` module to just give me implicit defs which would generate `AsyncHandler` instances for types with `Sync` or `Async` instances, hence this PR.

**Comment by andimiller** [comment]
I worked out that I can keep the previous behaviour by using a low priority trait for the `implicit def`

**Comment by gmethvin** [comment]
@sksamuel we'll need to do a new release for this

**Comment by sksamuel** [comment]
Yep, I want to add a ZIO module tomorrow then I'll release.

**Pull request merged**